docs/developer-docs/security/rust-canister-development-security-best-practices.md

```diff
@@ -10,20 +10,32 @@ This document contains canister development best practices for both Motoko and R
 
 #### Security concern
 
-The controller of a canister can change or update the canister whenever they like. If a canister e.g. stores assets such as ICP, this effectively means that the controller can steal these by updating the canister and transfer the cycles to their account.
+The controller of centralized components of a dapp (such as centrally controlled canisters or cloud services) can change or update these components whenever they like. If a canister or off-chain component e.g. holds assets such as ICP, ckBTC or ckETH on the user's behalf, this effectively means that the controller could decide at any time to steal these funds by e.g. updating the canister and transferring the assets to their account.
 
-#### Recommendation
+Furthermore, the controller of canisters serving web content (such as e.g. the [asset canister](../../references/asset-canister)) could maliciously modify the web application to e.g. steal user funds or perform security sensitive actions on the user's behalf. For example, if [Internet Identity](../../developer-docs/integrations/internet-identity/overview) is used, the user principal's private key for the given origin is stored in the browser storage, and a malicious app can therefore fully control the private key and the user's session and any assets controlled by that key.  
+
+#### Recommendations
+
+The following list starts with recommendations for dapps under centralized control, and then provides increasingly stronger recommendations to strengthen the dapp's security in terms of decentralization. 
+1. **The dapp uses central, off-chain components:** The application makes use of centralized components that are e.g. running in the cloud. The owners of these cloud services have full control over the application and assets managed by it. Your application should likely be further decentralized by avoiding central components. But while you have them, [securely manage your keys in the cloud](https://cloudsecurityalliance.org/research/topics/cloud-key-management/). 
+2. **Centralized dapp control by the developer team:** Your project is not controlled by a DAO, for example because it is in an early development stage or does not (yet) hold significant funds. In that case, we recommend managing access to your canisters securely and ideally not letting individuals take full control of the application. To achieve that, consider requiring several individuals or parties to control canisters and to make security sensitive changes on the application level, such as modifying access controls, etc. A helpful tool to achieve this is the [threshold canister](https://github.com/dfinity/threshold). Manage key material using hardware security modules, such as e.g. [YubyHSM](https://www.yubico.com/ch/store/yubihsm-2-series/) and physically protect these. Some of these devices support threshold signature schemes, which can help to further secure the setup. 
```

Comment:
```suggestion
2. **Centralized dapp control by the developer team:** Your project is not controlled by a DAO, for example because it is in an early development stage or does not (yet) hold significant funds. In that case, it is recommended to manage access to your canisters securely and ideally not letting individuals take full control of the application. To achieve that, consider requiring several individuals or parties to control canisters and to make security sensitive changes on the application level, such as modifying access controls, etc. A helpful tool to achieve this is the [threshold canister](https://github.com/dfinity/threshold). Manage key material using hardware security modules, such as e.g. [YubyHSM](https://www.yubico.com/ch/store/yubihsm-2-series/) and physically protect these. Some of these devices support threshold signature schemes, which can help to further secure the setup. 
```

I was not sure whether to read 
"consider requiring several individuals or parties (to control canisters and to make security sensitive changes on the application level)" or "(consider requiring several individuals or parties to control canisters) and to make security sensitive changes on the application level" and I was confused about what it means to "make security sensitive changes on the appliation level. Maybe this sentence can be simplified or split up?

Reply:
Can also be resolved from my POV, thanks!